turbopack/crates/turbo-tasks/src/parallel.rs

```diff
@@ -0,0 +1,367 @@
+//! Parallel for each resp. map running in the current tokio thread pool maintaining turbo tasks and
+//! tracing context.
+//!
+//! This avoid the problem of sleeping threads with mimalloc when using rayon in combination with
+//! tokio. It also avoid having multiple thread pools.
+
+use std::{
+    mem::{ManuallyDrop, transmute},
+    panic,
+    sync::{Arc, LazyLock},
+    thread::available_parallelism,
+};
+
+use tokio::{
+    runtime::Handle,
+    task::{JoinHandle, block_in_place},
+};
+use tracing::{Instrument, Span};
+
+use crate::{TurboTasksApi, turbo_tasks, turbo_tasks_scope};
+
+/// Calculates a good chunk size for parallel processing based on the number of available threads.
+/// This is used to ensure that the workload is evenly distributed across the threads.
+fn good_chunk_size(len: usize) -> usize {
+    static GOOD_CHUNK_COUNT: LazyLock<usize> =
+        LazyLock::new(|| available_parallelism().map_or(16, |c| c.get() * 4));
+    let min_chunk_count = *GOOD_CHUNK_COUNT;
+    (len + min_chunk_count - 1) / min_chunk_count
+}
+
+/// Context to allow spawning a task with a limited lifetime.
+struct ProcessInParallelContext<'l> {
+    handle: Handle,
+    turbo_tasks: Arc<dyn TurboTasksApi>,
+    span: Span,
+    phantom: std::marker::PhantomData<&'l ()>,
+}
+
+impl<'l> ProcessInParallelContext<'l> {
+    fn task<R, F>(&self, f: F) -> JoinHandle<R>
+    where
+        R: Send + 'static,
+        F: FnOnce() -> R + Send + 'l,
+    {
+        let f: Box<dyn FnOnce() -> R + Send + 'l> = Box::new(f);
+        let f: Box<dyn FnOnce() -> R + Send + 'static> = unsafe {
```

Comment:
add a safety comment

given that we return a joinhandle, im guessing this iis actually risky if the closure is dropped before the handle is joined

Reply:
Yes. The process_in_parallel is trying to prevent that.